webrender/src/device/gfx.rs

```diff
@@ -777,119 +777,155 @@ impl<B: hal::Backend> BufferPool<B> {
+    current_buffer_index: usize,
 }
 
 impl<B: hal::Backend> InstanceBufferHandler<B> {
     pub fn new(
         device: &B::Device,
-        memory_types: &Vec<hal::MemoryType>,
+        memory_types: &[hal::MemoryType],
         usage: hal::buffer::Usage,
         data_stride: usize,
         non_coherent_atom_size: usize,
         pitch_alignment_mask: usize,
     ) -> Self {
-        let mut buffer_pool = SmallVec::new();
-        buffer_pool.push(Buffer::new(
-            device,
-            memory_types,
-            usage,
-            pitch_alignment_mask,
-            MAX_INSTANCE_COUNT,
-            data_stride,
-        ));
+        let buffers = vec![
+            InstancePoolBuffer::new(
+                device,
+                memory_types,
+                usage,
+                data_stride,
+                pitch_alignment_mask,
+            )];
+
         InstanceBufferHandler {
-            offset: SmallVec::from_vec(vec![0]),
-            size: SmallVec::from_vec(vec![0]),
+            buffers,
+            memory_types: memory_types.to_vec(),
             non_coherent_atom_size,
             pitch_alignment_mask,
-            buffer_pool,
             data_stride,
-            buffer_index: 0,
+            current_buffer_index: 0,
         }
     }
 
     pub fn add<T: Copy>(
         &mut self,
         device: &B::Device,
         data: &[T],
-        memory_types: &Vec<hal::MemoryType>,
     ) {
-        let data_len = data.len();
-        let buffer_len = data_len * self.data_stride;
-        if self.offset() * self.data_stride + buffer_len > self.buffer().buffer_size {
-            let update_len = (self.buffer().buffer_size / self.data_stride - self.offset()) &! (mem::size_of::<T>() - 1);
-            if update_len > 0 {
-                self.buffer().update(
-                    device,
-                    &data[0 .. update_len],
-                    self.offset(),
-                    self.non_coherent_atom_size,
-                );
-                self.size[self.buffer_index] = update_len;
-                self.offset[self.buffer_index] += update_len;
-            }
-            let data = &data[update_len ..];
-            self.buffer_index += 1;
-            self.offset.push(0);
-            self.size.push(0);
-            if self.buffer_pool.len() < (self.buffer_index + 1) {
-                self.buffer_pool.push(
-                    Buffer::new(
-                        device,
-                        memory_types,
-                        hal::buffer::Usage::VERTEX,
-                        self.pitch_alignment_mask,
-                        MAX_INSTANCE_COUNT,
-                        self.data_stride,
+        let mut data = data;
+        while data.len() > 0 {
+            if self.current_buffer().buffer.buffer_size - self.current_buffer().offset * self.data_stride < mem::size_of::<T>() {
+                self.current_buffer_index += 1;
+                if self.buffers.len() < self.buffer_count() {
```

Comment:
what happens if we reach it?

Reply:
If we reach it we don't have to allocate new buffer, since we already have enough.